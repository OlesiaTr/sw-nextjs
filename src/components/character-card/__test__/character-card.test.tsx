import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import CharacterCard from '../character-card';
import type { Character } from '@/helpers';

describe('CharacterCard', () => {
  const MOCK_FUNCTION = jest.fn();
  const MOCK_CHARACTER: Character = {
    name: 'Luke Skywalker',
    gender: 'male',
    birth_year: '19BBY',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    homeworld: 20,
    id: 1,
    created: '2014-12-10T16:16:29.192000Z',
    edited: '2014-12-20T21:17:50.325000Z',
    films: [1, 2, 3, 4, 5, 6],
    species: [1],
    starships: [48, 59, 64, 65, 74],
    url: 'https://sw-api.starnavi.io/people/1/',
    vehicles: [38],
  };

  const renderCharacterCard = () =>
    render(
      <CharacterCard
        character={MOCK_CHARACTER}
        onCharacterClick={MOCK_FUNCTION}
      />,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders character details correctly', () => {
    const { asFragment } = renderCharacterCard();

    expect(screen.getByText(MOCK_CHARACTER.name)).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.gender)).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.birth_year)).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.height + ' cm')).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.mass + ' kg')).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.hair_color)).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.skin_color)).toBeInTheDocument();
    expect(screen.getByText(MOCK_CHARACTER.eye_color)).toBeInTheDocument();
    expect(
      screen.getByText('Homeworld: ' + MOCK_CHARACTER.homeworld),
    ).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onCharacterClick with the correct id when clicked', () => {
    renderCharacterCard();

    const card = screen.getByTestId(
      `character-card-${MOCK_CHARACTER.id}-test-id`,
    );

    if (card) {
      fireEvent.click(card);
      expect(MOCK_FUNCTION).toHaveBeenCalledWith(MOCK_CHARACTER.id);
    }
  });
});
