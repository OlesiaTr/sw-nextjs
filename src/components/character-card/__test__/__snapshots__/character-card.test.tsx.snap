// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard renders character details correctly 1`] = `
<DocumentFragment>
  <li
    class="flex w-[400px] cursor-pointer flex-col justify-between rounded-xl bg-white p-5 shadow transition hover:scale-95 hover:bg-custom-gradient"
    data-testid="character-card-1-test-id"
  >
    <div
      class="mb-3 flex justify-between"
    >
      <h3
        class="text-2xl font-black uppercase text-accent"
      >
        Luke Skywalker
      </h3>
      <p
        class="text-lg font-medium uppercase"
      >
        male
      </p>
    </div>
    <div
      class="rounded-sm border-borderGray bg-mainGray px-3 py-1"
    >
      <div
        class="mb-3 flex flex-wrap justify-between font-medium"
      >
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Birth Year:
          </span>
           19BBY
        </p>
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Height:
          </span>
           172 cm
        </p>
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Mass:
          </span>
           77 kg
        </p>
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Hair Color:
          </span>
           blond
        </p>
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Skin Color:
          </span>
           fair
        </p>
        <p
          class="mb-2 w-1/2"
        >
          <span
            class="font-semibold"
          >
            Eye Color:
          </span>
           blue
        </p>
      </div>
      <div>
        <p
          class="text-2xl font-bold"
        >
          Homeworld: 20
        </p>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
